### DISCUSSION TITLE
`SAWarning: nested transaction already deassociated from connection` migrating from SQLAlchemy 1.4 to 2

### DISCUSSION THREAD

**Discussion by cjw296** [sqlalchemy/sqlalchemy#12006] [Usage Questions] [answered]
I'm migrating [a library](https://github.com/Mortar/mortar_sqlalchemy) from SQLalchemy 1.4 to 2, but I've hit a snag with my [nested_transaction_on_connection](https://github.com/Mortar/mortar_sqlalchemy/blob/eb9cdfa8613a7a0c0775abbbe7c634bd311dd577/mortar_sqlalchemy/testing.py#L29-L36) testing helper.

I believe this is a minimal reproducer, with the problematic code just by the comment about flush issuing rollback

```python
from contextlib import contextmanager
from os import environ
from sqlalchemy import Column, Integer, create_engine
from sqlalchemy.orm import declarative_base, Session
Base = declarative_base()

class Model(Base):
    __tablename__ = 'model'
    id = Column(Integer, primary_key=True)
    value = Column(Integer)

@contextmanager
def connection_in_transaction():
    engine = create_engine(environ['DB_URL'])
    conn = engine.connect()
    transaction = conn.begin()
    try:
        yield conn
    finally:
        transaction.rollback()

@contextmanager
def nested_transaction_on_connection(conn):
    transaction = conn.begin_nested()
    try:
        yield
    finally:
        # https://docs.sqlalchemy.org/en/14/faq/sessions.html#but-why-does-flush-insist-on-issuing-a-rollback
        if transaction.is_active:
            transaction.rollback()

@contextmanager
def create_tables_and_session(conn):
    with nested_transaction_on_connection(conn):
        Base.metadata.create_all(bind=conn, checkfirst=False)
        yield Session(conn)


def test_nested_transaction_reproducer():
    with connection_in_transaction() as conn:
        with create_tables_and_session(conn) as session:
            session.add(Model(id=1, value=1))
            session.flush()
```

When I run `pytest -W 'error:nested transaction already' tests/test_common.py::test_nested_transaction_reproducer`, I get:

```

    def test_nested_transaction_reproducer():
        with connection_in_transaction() as conn:
>           with create_tables_and_session(conn) as session:

tests/test_common.py:128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py:144: in __exit__
    next(self.gen)
tests/test_common.py:121: in create_tables_and_session
    with nested_transaction_on_connection(conn):
/Library/Frameworks/Python.framework/Versions/3.12/lib/python3.12/contextlib.py:144: in __exit__
    next(self.gen)
tests/test_common.py:117: in nested_transaction_on_connection
    transaction.rollback()
../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/engine/base.py:2610: in rollback
    self._do_rollback()
../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/engine/base.py:2851: in _do_rollback
    self._close_impl(True, True)
../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/engine/base.py:2841: in _close_impl
    self._deactivate_from_connection(warn=warn_already_deactive)
../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/engine/base.py:2810: in _deactivate_from_connection
    util.warn(
../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py:1817: in warn
    _warnings_warn(msg, exc.SAWarning)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = 'nested transaction already deassociated from connection', category = <class 'sqlalchemy.exc.SAWarning'>, stacklevel = 6

    def _warnings_warn(
        message: Union[str, Warning],
        category: Optional[Type[Warning]] = None,
        stacklevel: int = 2,
    ) -> None:
        # adjust the given stacklevel to be outside of SQLAlchemy
        try:
            frame = sys._getframe(stacklevel)
        except ValueError:
            # being called from less than 3 (or given) stacklevels, weird,
            # but don't crash
            stacklevel = 0
        except:
            # _getframe() doesn't work, weird interpreter issue, weird,
            # ok, but don't crash
            stacklevel = 0
        else:
            stacklevel_found = warning_tag_found = False
            while frame is not None:
                # using __name__ here requires that we have __name__ in the
                # __globals__ of the decorated string functions we make also.
                # we generate this using {"__name__": fn.__module__}
                if not stacklevel_found and not re.match(
                    _not_sa_pattern, frame.f_globals.get("__name__", "")
                ):
                    # stop incrementing stack level if an out-of-SQLA line
                    # were found.
                    stacklevel_found = True
    
                    # however, for the warning tag thing, we have to keep
                    # scanning up the whole traceback
    
                if frame.f_code in _warning_tags:
                    warning_tag_found = True
                    (_suffix, _category) = _warning_tags[frame.f_code]
                    category = category or _category
                    message = f"{message} ({_suffix})"
    
                frame = frame.f_back  # type: ignore[assignment]
    
                if not stacklevel_found:
                    stacklevel += 1
                elif stacklevel_found and warning_tag_found:
                    break
    
        if category is not None:
>           warnings.warn(message, category, stacklevel=stacklevel + 1)
E           sqlalchemy.exc.SAWarning: nested transaction already deassociated from connection

../../../virtualenvs/mortar_sqlalchemy_312/lib/python3.12/site-packages/sqlalchemy/util/langhelpers.py:1892: SAWarning
```

If I comment out the `flush()`, then the warning goes away.

This [works fine](https://app.circleci.com/pipelines/github/Mortar/mortar_sqlalchemy/104/workflows/8dcdb488-f84f-4dcf-876a-fda47e87de5b/jobs/539/parallel-runs/0/steps/0-104) under under SQLAlchemy 1.4.54 (note the lack of similar warning in the pytest output), so what should I be doing to get the same result in SQLAlchemy 2?

**Comment by CaselIT** [comment]
hi,

>  SQLalchemy 2 to 3,

you mean 1.4 to 2? because there is no 3

> I believe this is a minimal reproduce

what is db?

**Reply by cjw296** [reply]
Yes, 1.4 to 2, Just corrected.

db is a fixture in the library, not actually needed here, it's the `conn` returned by `connection_in_transaction`, so I've updated the example, thanks for spotting these two!

**Reply by CaselIT** [reply]
ok, so you likely need to play with https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode.

the default sees that the connection is in a nested transaction, so the session will use nested transactions. When the session is closed the nested transaction is likely closed that leads to the warning.

**Reply by cjw296** [reply]
Well no, the intention here is that the code under test doesn't change a t all.
The `is_active` check is what makes this work in 1.4, I'm curious as to whether the fact that it's `True` after the subtransaction has evidently been rolled back following the flush is a bug in 2. @zzzeek have me the original fix, so hopefully he'll feel guilty enough to help here 😆

**Reply by zzzeek** [reply]
@CaselIT isn't wrong here, you might want to look at the transaction mode stuff.    I'm pretty unavailable today but I can try to have a look later

**Reply by cjw296** [reply]
Okay, this this does work:

```python
        yield Session(db, join_transaction_mode="control_fully")
```

...to make it like it was in 1.4. Cool, that gets me over this hump, but I feel I will be back with a variant of this question some time soon 😓

**Reply by CaselIT** [reply]
you may find this useful https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites

**Reply by cjw296** [reply]
I tried that, but I have:

- a session scoped pytest fixture that creates the database, enables postgres extensions, etc. This is done inside a connection-level transaction

- a per-test scoped pytest fixture that, in a nested transaction, creates any tables and the ORM Session which is then handed to the test that asked for it

If I use `join_transaction_mode="create_savepoint"` the warnings come back from the nested transaction closing. If I try and remove the nested transaction, then obviously the tables and rows created in each test aren't cleaned up between tests as the session scoped fixture's teardown won't have run...

**Reply by CaselIT** [reply]
this discussion seems somewhat similar https://github.com/sqlalchemy/sqlalchemy/discussions/11795

maybe you can look if the example there can work in your case?

**Comment by zzzeek** [comment]
I dont understand the purpose of placing the connection into a savepoint before passing it to the Session.     Here is the correct way to use the pattern:  https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites

1. put the connection into a transaction (but not a savepoint)
2. use the "create savepoint" join transaction mode

this is a. less code for you and b. we test and support this pattern


```py
@contextmanager
def connection_in_transaction():
    engine = create_engine("sqlite://", echo=True)
    conn = engine.connect()
    transaction = conn.begin()
    try:
        yield conn
    finally:
        transaction.rollback()

@contextmanager
def create_tables_and_session(conn):
    Base.metadata.create_all(bind=conn, checkfirst=False)
    yield Session(conn, join_transaction_mode="create_savepoint")


def test_nested_transaction_reproducer():
    with connection_in_transaction() as conn:
        with create_tables_and_session(conn) as session:
            session.add(Model(id=1, value=1))
            session.flush()
```

**Reply by cjw296** [reply]
@CaselIT is spot on that https://github.com/sqlalchemy/sqlalchemy/discussions/11795 is very close to what I'm after.

So, in this case, my original intention was what I said in [this comment](https://github.com/sqlalchemy/sqlalchemy/discussions/12006#discussioncomment-10990180).

But, for this library, the expensive bit is actually creating the database (which can mean spinning up a container, creating a database, etc). Separating _that_ out into the session-scoped fixture means for this library the pattern you describe will work, so I'll use that...

Done in https://github.com/Mortar/mortar_sqlalchemy/commit/627e607c7e11771138e4dc9c8bde5f2e7bb12ab3

...however, I can certainly see a case where I want to:

- for each pytest session, create a database, potentially with a big schema probably set up with an alembic run
- for each pytest session, or maybe a module or package level set of tests, insert a bunch of data for for tests that need a big, common set, which I'd assume I'd need to do in a savepoint/subtransaction
- run each test in a further savepoint/subtransaction so it can be quickly and cheaply roll back any impact the test has had.

Are there documented patterns for this?
What do you recommend?

**Reply by zzzeek** [reply]
I would refer to my answer at https://github.com/sqlalchemy/sqlalchemy/discussions/11795#discussioncomment-10476235

"run each test in a savepoint"

I would **not** do that.  tests should be run in **real transactions**.  you do not want transactional effects leaking between tests.

as stated in my comment, "insert a bunch of data" is at the **same level** as DDL.   DDL is also transactional.   the CREATE and INSERT statements used by all your tests in a common way are at the same level.

otherwise you're making the same mistake they are in #11795 , fixed test data that is common to all tests being treated as somehow more special than the DDL structures themselves.

**Reply by cjw296** [reply]
Okay, so in the event of: 

- a complex and large schema
- expensive-to-insert sets of data that are applicable to a group of tests
- the desire to run each test in isolation
- pytest as the runner
- postgres only, to make things easier ;-)

What fixtures do you recommend and what transaction  structure would they result in?

**Reply by zzzeek** [reply, marked as answer]
depends on what changes per group of tests but lets propose:

1. suite-level fixture that does table DDL and create schema, drop schema.  This can also do INSERTs of data that you think apply to all tests globally
3. if you have sets of data applicable to groups of tests but not others, then a module-level fixture that does INSERTs local to group of tests, does a COMMIT, on cleanup does DELETE from all tables if you want that data to be transitory for groups of tests
4. test level fixtures that do transactions using the pattern at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites  (this is trivial to adapt to pytest)


now if you do actually want savepoint-per-test  and transaction-per-group-of-tests, that should be possible also anyway, the pattern at https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites can be given a SAVEPOINT as well.  Here's a full POC, two tests, each modify the table locally, includes global data in the table with just one INSERT

```py
import pytest

from sqlalchemy import create_engine
from sqlalchemy import insert
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class A(Base):
    __tablename__ = "a"

    id: Mapped[int] = mapped_column(primary_key=True)
    data: Mapped[str]


@pytest.fixture(scope="module")
def engine():
    e = create_engine("sqlite://", echo=True)
    yield e
    e.dispose()


@pytest.fixture(scope="module")
def tables(engine):
    with engine.begin() as conn:
        Base.metadata.create_all(conn)
    yield
    with engine.begin() as conn:
        Base.metadata.drop_all(conn)


@pytest.fixture(scope="module")
def fixture_data(engine):
    conn = engine.connect()
    trans = conn.begin()
    conn.execute(insert(A), [{"data": f"d{i}"} for i in range(10)])
    yield conn, trans
    conn.rollback()


@pytest.fixture(scope="function")
def session(engine, tables, fixture_data):
    connection, trans = fixture_data

    savepoint = connection.begin_nested()

    # bind an individual Session to the connection, selecting
    # "create_savepoint" join_transaction_mode
    session = Session(
        bind=connection, join_transaction_mode="create_savepoint"
    )
    yield session
    session.close()

    savepoint.rollback()


def test_one(session):
    session.execute(insert(A), {"data": "test_one_only"})
    assert set(session.scalars(select(A.data))) == {"test_one_only"}.union(
        f"d{i}" for i in range(10)
    )


def test_two(session):
    session.execute(insert(A), {"data": "test_two_only"})
    assert set(session.scalars(select(A.data))) == {"test_two_only"}.union(
        f"d{i}" for i in range(10)
    )

```


SQL output

```
platform linux -- Python 3.12.4, pytest-8.1.0, pluggy-1.4.0 -- /home/classic/.venv3/bin/python
cachedir: .pytest_cache
rootdir: /home/classic/dev/sqlalchemy
configfile: pyproject.toml
plugins: xdist-3.4.0, anyio-4.1.0, random-0.2, repeat-0.9.3
collected 2 items                                                                                                                                                            

test3.py::test_one 2024-10-22 14:51:18,420 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("a")
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("a")
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine 
CREATE TABLE a (
	id INTEGER NOT NULL, 
	data VARCHAR NOT NULL, 
	PRIMARY KEY (id)
)


2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine COMMIT
2024-10-22 14:51:18,421 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-10-22 14:51:18,422 INFO sqlalchemy.engine.Engine INSERT INTO a (data) VALUES (?)
2024-10-22 14:51:18,422 INFO sqlalchemy.engine.Engine [generated in 0.00008s] [('d0',), ('d1',), ('d2',), ('d3',), ('d4',), ('d5',), ('d6',), ('d7',), ('d8',), ('d9',)]
2024-10-22 14:51:18,422 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_1
2024-10-22 14:51:18,422 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,423 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_2
2024-10-22 14:51:18,423 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,424 INFO sqlalchemy.engine.Engine INSERT INTO a (data) VALUES (?)
2024-10-22 14:51:18,424 INFO sqlalchemy.engine.Engine [generated in 0.00007s] ('test_one_only',)
2024-10-22 14:51:18,424 INFO sqlalchemy.engine.Engine SELECT a.data 
FROM a
2024-10-22 14:51:18,424 INFO sqlalchemy.engine.Engine [generated in 0.00007s] ()
PASSED2024-10-22 14:51:18,425 INFO sqlalchemy.engine.Engine ROLLBACK TO SAVEPOINT sa_savepoint_2
2024-10-22 14:51:18,425 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,425 INFO sqlalchemy.engine.Engine ROLLBACK TO SAVEPOINT sa_savepoint_1
2024-10-22 14:51:18,425 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()

test3.py::test_two 2024-10-22 14:51:18,425 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_3
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine SAVEPOINT sa_savepoint_4
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine [no key 0.00006s] ()
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine INSERT INTO a (data) VALUES (?)
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine [cached since 0.002424s ago] ('test_two_only',)
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine SELECT a.data 
FROM a
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine [cached since 0.001769s ago] ()
PASSED2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine ROLLBACK TO SAVEPOINT sa_savepoint_4
2024-10-22 14:51:18,426 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine ROLLBACK TO SAVEPOINT sa_savepoint_3
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine [no key 0.00005s] ()
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine ROLLBACK
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("a")
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine 
DROP TABLE a
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine [no key 0.00004s] ()
2024-10-22 14:51:18,427 INFO sqlalchemy.engine.Engine COMMIT


============================================================================= 2 passed in 0.40s ==============================================================================
```

**Reply by CaselIT** [reply]
on 2. I suggest truncate instead of delete without where, since it's usually a lot faster (just unlink the data files :) )

**Comment by cjw296** [comment]
Just following up that I just bumped into my previous question on this from a couple of years before, same warning anyway, I was trying to find this discussion so might as well link them for future travellers:

https://github.com/sqlalchemy/sqlalchemy/discussions/8621
